### DISCUSSION TITLE
How to add shared state to Loco?

### DISCUSSION THREAD

**Discussion by georgeboot** [loco-rs/loco#469] [Q&A]
I recently started playing with Loco and an very impressed. We use Laravel (PHP) but needs better performance for our data operations (converting images, transforming coordinate systems, processing large pointclouds etc).

While experimenting with [libvips](https://crates.io/crates/libvips) for image manipulation, I noticed that the crate depends on a "context":
```rs
let app = VipsApp::new("Test Libvips", false).expect("Cannot initialize libvips");
``` 

What is the best way in Loco for this variable to be available everywhere?

My initial though was to use the `after_routes` hook like so:
```rs
async fn after_routes(router: Router, _ctx: &AppContext) -> Result<Router> {
    let vips_context = VipsApp::new("Test Libvips", false).expect("Cannot initialize libvips");
    vips_context.concurrency_set(8);

    let vips_context_wrapped = Arc::new(RwLock::new(vips_context));

    Ok(router.layer(Extension(vips_context_wrapped)))
}
```

but I believe it's then only available in routes, and not in, say, taks or workers.

**Comment by schungx** [comment]
I suppose it is best to build an "initializer" which essentially is a shared state that you can then call up in your request.

It is quite simple to build one actually...  Unfortunately the docs does not have step-by-step instructions.

However look into the view-initializer implementation.

**Reply by georgeboot** [reply]
That seems to be only shared between requests (routes), and is not accessibly from tasks or workers.

Maybe what im asking is not possible yet, but for example the database connection pool does a similar thing.

**Comment by kaplanelad** [comment]
Hey @georgeboot,

As @schungx mentioned, the method to share state involves utilizing `after_routes` and incorporating the Axum Extension layer. You're correct; this happens between requests.

Could you provide more details on what you're aiming to achieve?

**Comment by mstallmo** [comment]
I think what @georgeboot is asking here relates to #522 where the internal `AppContext` is only state that is shared to tasks and workers.

If you wanted to implement a background task that used the shared libvips context you wouldn't be able to. Instead you would have to create a new libvips context in the `perform` method of the workers.

**Comment by jondot** [comment]
This is a great question (and I believe should be worth a section in the docs).

**Shared state in controllers:**
You're correct. Use Axum infra for this, like in here: https://github.com/loco-rs/loco/blob/master/examples/llm-candle-inference/src/app.rs#L41


**Shared state in workers:**
Workers are intentionally _verbatim_ initialized in app hooks:
https://github.com/loco-rs/loco/blob/master/starters/saas/src/app.rs#L59
This means you can shape them as a "regular" Rust struct that takes a state as a field. Then refer to that field in `perform`.


Note that by-design sharing state *between controllers and workers* have no meaning, because even though you may choose to run workers in the same process as controllers initially (and share state) -- you'd want to quickly switch to proper workers backed by queue and running in a standalone workers process as you scale horizontally, and so workers should by-design have no shared state with controllers, for your own good.

**Tasks**
Tasks don't really have a value for shared state, as they have a similar life as any exec'd binary. The process fires up, boots, creates all resources needed (connects to db, etc.), performs the task logic, and then the process is dead.
And so, you can just create what ever state or utility you need as the task runs.

**Comment by mstallmo** [comment]
I think it would be great to expand on the philosophy behind this in the docs just to add some clarity!

IMO where this difference gets a bit confusing is in two main places.

1) Coming from other frameworks like Rails you generally have access to the global application state in workers as well so the lack of a similar pattern can be unexpected. 
2) The fact that `AppContext` is passed into the worker's `build` method and is present in controllers suggests that application state is, and should be, shared between controllers so it follows that other state aside from `AppContext` could be shared as well.

Granted, even in Rails, the state isn't truly shared between workers running on a dedicated machine and the main application server itself but it gives the illusion of this via the whole app being booted to run the worker jobs. So anything that is available via things like initializers in the main application server is also available in workers.

**Reply by jondot** [reply]
I think the difference here is serializable state vs non-serializable.

In the rails case, the illusion is created by hydrating back the state so that workers (living on different processes, or machines) refer to it as just any other field.

For the specific vips example, looks like the requirement is a hard one -- having a single dynamic library loaded into memory per process, so it's almost like having to throw in a mutex over it and use some classic concurrency tools to make it work (this would have been correct in any other programming language).

Either way -- I've created a live example [in this repo](https://github.com/loco-rs/shared-global-state) using libvips, and added a section discussing this in the docs.

Thanks to all bringing up this interesting use case!
